Make things a bit more OO

benbalter opened the pull request:
This pull request started out as an attempt to add a `Repository` class, that could accept a repository name and retrieve the repository's CNAME, but expanded to abstract much of the check logic out of a single giant class, into more object-oriented classes. Specifically:

* Move domain-specific checks to the `Domain` class
* Add new repo-level checks to a `Repository` class
* Create a new `Site` class to encompass `Domain` and `Repository` checks
* Abstract `check!`, `valid?` and `reason` logic to a `Checkable` class
* Create an `Errors` class to autoload all errors
* Make `GitHubPages` and `HealthCheck` modules, not classes

Still to do:

- [x] Tests

parkr commented on lib/github-pages-health-check.rb:
```diff
@@ -8,19 +8,25 @@
 
-class GitHubPages
-  class HealthCheck
+if File.exists?(File.expand_path ".env", File.dirname(__FILE__))
```
You'd vendor a `.env` inside of `lib/`?

parkr commented on lib/github-pages-health-check.rb:
```diff
@@ -8,19 +8,25 @@
+    autoload :Repository, "github-pages-health-check/repository"
+    autoload :Site,       "github-pages-health-check/site"
 
     LEGACY_IP_ADDRESSES = %w[
```
Should these go in a `Checkable` that implements IP checking?

parkr commented on lib/github-pages-health-check/cloudflare.rb:
```diff
@@ -1,9 +1,10 @@
-class GitHubPages
-  class HealthCheck
+module GitHubPages
```
This might collide with other `GitHubPages` modules or classes... I think it's a class in some places (like shell)?

benbalter replied:
Which is the "right" way to do it? Figure it's easy to update one-off scripts. Could also be `GitHub::Pages::HealthCheck`?

parkr replied:
True. I guess if you don't think there's ever a reason you should have to instantiate `GitHubPages`, then it should be a module. :)

parkr commented on lib/github-pages-health-check/domain.rb:
```diff
@@ -0,0 +1,202 @@
+module GitHubPages
+  module HealthCheck
+    class Domain < GitHubPages::HealthCheck::Checkable
```
You probably only need `Checkable` here, but more a style thing than anything about correctness.

parkr commented on lib/github-pages-health-check/domain.rb:
```diff
@@ -0,0 +1,202 @@
+
+      def deprecated_ip?
+        return @deprecated_ip if defined? @deprecated_ip
+        @deprecated_ip = a_record? && old_ip_address?
```
This leaves a bit of ambiguity and I'm not sure how Ruby's lexer deals with it. Parentheses around `a_record? && old_ip_address?` might help?

parkr commented on lib/github-pages-health-check/domain.rb:
```diff
@@ -0,0 +1,202 @@
+        @invalid_cname = begin
+          return false unless valid_domain?
+          !github_domain? && !apex_domain? && !pointed_to_github_user_domain?
+        end
```
This could be simplified to

```ruby
@invalid_cname = !valid_domain? || !(github_domain? || apex_domain? || pointed_to_github_user_domain?)
```

I flipped the `!... && !..` to `!(... || ...)` for slightly easier reading. Feel free to use `(!.... && !....)` if it's easier to read the other way.

parkr commented on lib/github-pages-health-check/domain.rb:
```diff
@@ -0,0 +1,202 @@
+      attr_reader :host
+
+      def initialize(host)
+        @host = host_from_uri(host)
```
It might be good to defensively check the input here:

```
raise ArgumentError, "Expected string, got #{host.class}" unless host.is_a? String
```

I see a lot of `undefined method for NilClass:nil` in our future without it :)

parkr commented on lib/github-pages-health-check/repository.rb:
```diff
@@ -0,0 +1,70 @@
+
+      def initialize(name_with_owner, access_token: nil)
+        unless name_with_owner =~ REPO_REGEX
+          raise GitHubPages::HealthCheck::Errors::InvalidRepositoryError
```
For consistency with the rest of the class, this should either be `Errors::InvalidRepositoryError`, or all the others should have `GitHubPages::HealthCheck::` :smile:

parkr commented on lib/github-pages-health-check/site.rb:
```diff
@@ -0,0 +1,67 @@
+        hash = {}
+
+        DOMAIN_HASH_METHODS.each do |method|
+          hash[method] = domain.send(method)
```
I'd use `public_send` here just to be a bit safer.

An alternative is to implement `#[]` for `Repository`!

parkr commented on lib/github-pages-health-check/site.rb:
```diff
@@ -0,0 +1,67 @@
+      alias_method :as_json, :to_hash
+
+      def to_json
+        as_json.to_json
```
What is this convention about? What does `to_json` return? A string?

parkr commented on lib/github-pages-health-check/site.rb:
```diff
@@ -0,0 +1,67 @@
+      def to_s
+        to_hash.inject(Array.new) do |all, pair|
+          all.push pair.join(": ")
+        end.join("\n")
```
We should get some justifications in here so it's easier to scan, and perhaps reorder the keys so the important info is at the top. Things like `valid?` should be first.

benbalter commented:
@parkr believe this should be ready for one more :eyes:, when you're ready. I believe I implemented everything you mentioned, plus tests, and simplified the `to_hash` logic by moving the methods that should become part of the hash logic to checkable.

parkr commented on README.md:
```diff
@@ -22,20 +22,20 @@
 => false
 > check.valid!
-=> GitHubPages::HealthCheck::InvalidCNAME
+=> GitHubPages::HealthCheck::Errors::InvalidCNAMEError
```
Does this return the class or raise an error?

benbalter replied:
Raises. What's the best way to note that?

parkr commented:
:shipit: 

Can we also ship #37.

benbalter commented:
> Can we also ship #37?

Looks like the conflict should be minimal. Once this merges, I can work on reconcile #37 and we can release a new version.
